nixos/modules/services/continuous-integration/gitlab-runner.nix

```diff
@@ -4,15 +4,60 @@ with lib;
 
 let
   cfg = config.services.gitlab-runner;
-  configFile = pkgs.writeText "config.toml" cfg.configText;
   hasDocker = config.virtualisation.docker.enable;
 in
 {
   options.services.gitlab-runner = {
     enable = mkEnableOption "Gitlab Runner";
 
-    configText = mkOption {
-      description = "Verbatim config.toml to use";
+    configFile = mkOption {
+      description = ''
+        Configuration file for gitlab-runner.
+        Use this option in favor of configOptions to avoid placing CI tokens in the nix store.
+
+        <option>configFile</option> takes precedence over <option>configOptions</option>.
+
+        Warning: Not using <option>configFile</option> will potentially result in secrets
+        leaking into the WORLD-READABLE nix store.
+      '';
+      type = types.path;
+      default = pkgs.runCommand "config.toml" {
+        buildInputs = [ pkgs.remarshal ];
+      } ''
+        remarshal -if json -of toml \
+          < ${pkgs.writeText "config.json" (builtins.toJSON cfg.configOptions)} \
+          > $out
+      '';
```

Comment:
This will rebuild the manual if `configOptions` is changed.

You could set `configFile` in the `config` section instead:

```services.gitlab-runner.configFile = mkIf (configOptions != null) (pkgs.runCommand "config.toml" ...);```

Also, if the user sets both `configFile` and `configOptions` this will automatically result in an error because `configFile` will be sett multiple times.